File: mifosng-android/src/test/java/com/mifos/mifosxdroid/online/GroupsListPresenterTest.java
package com.mifos.mifosxdroid.online;

import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.mifos.api.datamanager.DataManagerGroups;
import com.mifos.mifosxdroid.FakeRemoteDataSource;
import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.online.groupslist.GroupsListMvpView;
import com.mifos.mifosxdroid.online.groupslist.GroupsListPresenter;
import com.mifos.mifosxdroid.util.RxSchedulersOverrideRule;
import com.mifos.objects.client.Client;
import com.mifos.objects.client.Page;
import com.mifos.objects.group.Group;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import rx.Observable;

/**
 * Created by Rajan Maurya on 28/6/16.
 */
@RunWith(MockitoJUnitRunner.class)
public class GroupsListPresenterTest {

    @Rule
    public final RxSchedulersOverrideRule mOverrideSchedulersRule = new RxSchedulersOverrideRule();

    GroupsListPresenter mGroupsListPresenter;

    @Mock
    DataManagerGroups mDataManagerGroups;

    @Mock
    GroupsListMvpView mGroupsListMvpView;

    Page<Group> groupPage;

    int offset = 0;
    int limit = 100;

    @Before
    public void setUp() {

        mGroupsListPresenter = new GroupsListPresenter(mDataManagerGroups);
        mGroupsListPresenter.attachView(mGroupsListMvpView);

        groupPage = FakeRemoteDataSource.getGroups();

    }

    @After
    public void tearDown() {
        mGroupsListPresenter.detachView();
    }

    @Test
    public void testLoadGroup() {

        stubDatabaseGetAllGroups(Observable.just(groupPage));

        mGroupsListPresenter.loadDatabaseGroups();

        mGroupsListPresenter.setAlreadyClientSyncStatus();
        verify(mGroupsListMvpView, never()).showMessage(R.string.failed_to_load_db_groups);

        stubDataManagerGetGroups(Observable.just(groupPage));

        mGroupsListPresenter.loadGroups(false, offset);

        verify(mGroupsListMvpView).showProgressbar(true);
        verify(mGroupsListMvpView).showGroups(groupPage.getPageItems());
        verify(mGroupsListMvpView, never()).showFetchingError();
        verify(mGroupsListMvpView).showProgressbar(false);
    }

    @Test
    public void testLoadGroupFails() {

        stubDataManagerGetGroups(Observable.<Page<Client>>error(new RuntimeException()));

        mGroupsListPresenter.loadGroups(false, offset);
        verify(mGroupsListMvpView).showFetchingError();
        verify(mGroupsListMvpView, never()).showGroups(groupPage.getPageItems());
    }

    @Test
    public void testLoadMoreGroups() {

        stubDatabaseGetAllGroups(Observable.just(groupPage));

        mGroupsListPresenter.loadDatabaseGroups();

        mGroupsListPresenter.setAlreadyClientSyncStatus();
        verify(mGroupsListMvpView, never()).showMessage(R.string.failed_to_load_db_groups);

        stubDataManagerGetGroups(Observable.just(groupPage));

        mGroupsListPresenter.loadGroups(true, offset);

        verify(mGroupsListMvpView).showProgressbar(true);
        verify(mGroupsListMvpView).showLoadMoreGroups(groupPage.getPageItems());
        verify(mGroupsListMvpView, never()).showFetchingError();
        verify(mGroupsListMvpView).showProgressbar(false);
    }

    @Test
    public void testLoadMoreGroupFails() {

        stubDataManagerGetGroups(Observable.<Page<Client>>error(new RuntimeException()));

        mGroupsListPresenter.loadGroups(true, offset);
        verify(mGroupsListMvpView).showMessage(R.string.failed_to_fetch_groups);
        verify(mGroupsListMvpView, never()).showGroups(groupPage.getPageItems());
    }

    @Test
    public void testEmptyGroupList() {

        stubDataManagerGetGroups(Observable.just(new Page<Client>()));

        mGroupsListPresenter.loadGroups(false, offset);

        verify(mGroupsListMvpView).showProgressbar(true);
        verify(mGroupsListMvpView).showEmptyGroups(R.string.group);
        verify(mGroupsListMvpView).unregisterSwipeAndScrollListener();
        verify(mGroupsListMvpView, never()).showFetchingError();
        verify(mGroupsListMvpView).showProgressbar(false);
    }

    @Test
    public void testNoMoreGroupsAvailable() {

        stubDataManagerGetGroups(Observable.just(new Page<Client>()));

        mGroupsListPresenter.loadGroups(true, offset);

        verify(mGroupsListMvpView).showProgressbar(true);
        verify(mGroupsListMvpView).showMessage(R.string.no_more_groups_available);
        verify(mGroupsListMvpView, never()).showFetchingError();
        verify(mGroupsListMvpView).showProgressbar(false);
    }

    @Test
    public void testLoadDatabaseGroups() {
        stubDatabaseGetAllGroups(Observable.just(groupPage));

        mGroupsListPresenter.loadDatabaseGroups();
        mGroupsListPresenter.setAlreadyClientSyncStatus();
        verify(mGroupsListMvpView, never()).showMessage(R.string.failed_to_load_db_groups);
    }

    @Test
    public void testLoadDatabaseGroupsFails() {
        stubDatabaseGetAllGroups(Observable.<Page<Client>>error(new RuntimeException()));

        mGroupsListPresenter.loadDatabaseGroups();

        verify(mGroupsListMvpView).showMessage(R.string.failed_to_load_db_groups);
    }

    public void stubDataManagerGetGroups(Observable observable) {
        when(mDataManagerGroups.getGroups(true, offset, limit)).thenReturn(observable);
    }

    public void stubDatabaseGetAllGroups(Observable observable) {
        when(mDataManagerGroups.getDatabaseGroups()).thenReturn(observable);
    }

}

File: mifosng-android/src/test/java/com/mifos/mifosxdroid/online/GroupsListPresenterTest.kt
package com.mifos.mifosxdroid.online

import com.mifos.api.datamanager.DataManagerGroups
import com.mifos.mifosxdroid.FakeRemoteDataSource.groups
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.online.groupslist.GroupsListMvpView
import com.mifos.mifosxdroid.online.groupslist.GroupsListPresenter
import com.mifos.mifosxdroid.util.RxSchedulersOverrideRule
import com.mifos.objects.client.Page
import com.mifos.objects.group.Group
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.MockitoJUnitRunner
import rx.Observable

/**
 * Created by Rajan Maurya on 28/6/16.
 */
@RunWith(MockitoJUnitRunner::class)
class GroupsListPresenterTest {
    @get:Rule
    val mOverrideSchedulersRule = RxSchedulersOverrideRule()
    private lateinit var mGroupsListPresenter: GroupsListPresenter

    @Mock
    lateinit var mDataManagerGroups: DataManagerGroups

    @Mock
    lateinit var mGroupsListMvpView: GroupsListMvpView
    private lateinit var groupPage: Page<Group>
    var offset = 0
    var limit = 100

    @Before
    fun setUp() {
        mGroupsListPresenter = GroupsListPresenter(mDataManagerGroups)
        mGroupsListPresenter.attachView(mGroupsListMvpView)
        groupPage = groups
    }

    @After
    fun tearDown() {
        mGroupsListPresenter.detachView()
    }

    @Test
    fun testLoadGroup() {
        stubDatabaseGetAllGroups(Observable.just(groupPage))
        mGroupsListPresenter.loadDatabaseGroups()
        mGroupsListPresenter.setAlreadyClientSyncStatus()
        Mockito.verify(mGroupsListMvpView, Mockito.never())
            .showMessage(R.string.failed_to_load_db_groups)
        stubDataManagerGetGroups(Observable.just(groupPage))
        mGroupsListPresenter.loadGroups(false, offset)
        Mockito.verify(mGroupsListMvpView).showProgressbar(true)
        Mockito.verify(mGroupsListMvpView).showGroups(groupPage.pageItems)
        Mockito.verify(mGroupsListMvpView, Mockito.never()).showFetchingError()
        Mockito.verify(mGroupsListMvpView).showProgressbar(false)
    }

    @Test
    fun testLoadGroupFails() {
        stubDataManagerGetGroups(Observable.error(RuntimeException()))
        mGroupsListPresenter.loadGroups(false, offset)
        Mockito.verify(mGroupsListMvpView).showFetchingError()
        Mockito.verify(mGroupsListMvpView, Mockito.never()).showGroups(groupPage.pageItems)
    }

    @Test
    fun testLoadMoreGroups() {
        stubDatabaseGetAllGroups(Observable.just(groupPage))
        mGroupsListPresenter.loadDatabaseGroups()
        mGroupsListPresenter.setAlreadyClientSyncStatus()
        Mockito.verify(mGroupsListMvpView, Mockito.never())
            .showMessage(R.string.failed_to_load_db_groups)
        stubDataManagerGetGroups(Observable.just(groupPage))
        mGroupsListPresenter.loadGroups(true, offset)
        Mockito.verify(mGroupsListMvpView).showProgressbar(true)
        Mockito.verify(mGroupsListMvpView).showLoadMoreGroups(groupPage.pageItems)
        Mockito.verify(mGroupsListMvpView, Mockito.never()).showFetchingError()
        Mockito.verify(mGroupsListMvpView).showProgressbar(false)
    }

    @Test
    fun testLoadMoreGroupFails() {
        stubDataManagerGetGroups(Observable.error(RuntimeException()))
        mGroupsListPresenter.loadGroups(true, offset)
        Mockito.verify(mGroupsListMvpView).showMessage(R.string.failed_to_fetch_groups)
        Mockito.verify(mGroupsListMvpView, Mockito.never()).showGroups(groupPage.pageItems)
    }

    @Test
    fun testEmptyGroupList() {
        stubDataManagerGetGroups(Observable.just(Page<Group>()))
        mGroupsListPresenter.loadGroups(false, offset)
        Mockito.verify(mGroupsListMvpView).showProgressbar(true)
        Mockito.verify(mGroupsListMvpView).showEmptyGroups(R.string.group)
        Mockito.verify(mGroupsListMvpView).unregisterSwipeAndScrollListener()
        Mockito.verify(mGroupsListMvpView, Mockito.never()).showFetchingError()
        Mockito.verify(mGroupsListMvpView).showProgressbar(false)
    }

    @Test
    fun testNoMoreGroupsAvailable() {
        stubDataManagerGetGroups(Observable.just(Page<Group>()))
        mGroupsListPresenter.loadGroups(true, offset)
        Mockito.verify(mGroupsListMvpView).showProgressbar(true)
        Mockito.verify(mGroupsListMvpView).showMessage(R.string.no_more_groups_available)
        Mockito.verify(mGroupsListMvpView, Mockito.never()).showFetchingError()
        Mockito.verify(mGroupsListMvpView).showProgressbar(false)
    }

    @Test
    fun testLoadDatabaseGroups() {
        stubDatabaseGetAllGroups(Observable.just(groupPage))
        mGroupsListPresenter.loadDatabaseGroups()
        mGroupsListPresenter.setAlreadyClientSyncStatus()
        Mockito.verify(mGroupsListMvpView, Mockito.never())
            .showMessage(R.string.failed_to_load_db_groups)
    }

    @Test
    fun testLoadDatabaseGroupsFails() {
        stubDatabaseGetAllGroups(Observable.error(RuntimeException()))
        mGroupsListPresenter.loadDatabaseGroups()
        Mockito.verify(mGroupsListMvpView).showMessage(R.string.failed_to_load_db_groups)
    }

    private fun stubDataManagerGetGroups(observable: Observable<Page<Group>>) {
        Mockito.`when`(
            mDataManagerGroups.getGroups(true, offset, limit)
        ).thenReturn(observable)
    }

    private fun stubDatabaseGetAllGroups(observable: Observable<Page<Group>>) {
        Mockito.`when`(
            mDataManagerGroups.databaseGroups
        ).thenReturn(observable)
    }
}